### ISSUE TITLE
RuntimeError: CUDA error: device-side assert triggered while eval

### ISSUE THREAD

**Issue by avi-jain** [ThilinaRajapakse/simpletransformers#84]
**Describe the bug**
I trained a distilbert model for Classification. Now when I try to use the model to eval or predict I get a `RuntimeError: CUDA error: device-side assert triggered`
**To Reproduce**
Train a model using the params `ClassificationModel('distilbert', 'distilbert-base-uncased', num_labels=len(df_bert_train['labels'].unique()), args={'reprocess_input_data': True, 'overwrite_output_dir': True, 'max_seq_length': 64, 'train_batch_size':16, 'fp16':False, 'num_train_epochs': 10}) `

**Desktop (please complete the following information):**
 - Linux

**Comment by ThilinaRajapakse** [comment]
This is often caused by not setting the `num_labels` parameter to the same value that was used when training the model. Make sure that you use the same value when loading a saved model.

**Comment by avi-jain** [comment]
Additional context - I'm using eval in the same context (jupyter notebook) in which the model was trained. Not loading it again.
Although, by disabling GPUs and loading the model, eval works

**Comment by ThilinaRajapakse** [comment]
Can you show me the line you are executing to perform the evaluation? Also, the terminal where you launch the jupyter notebook might show some additional info about the error.

**Comment by avi-jain** [comment]
I checked the terminal output. Nothing except for a couple of warnings (WARNING | WARNING: attempted to send message from fork). I'll save all my transient variables, load the model after enabling gpus and try the eval again. 
The line is the same as your example 
`result, model_outputs, wrong_predictions = model.eval_model(df_bert_dev, verbose=True)
`

**Comment by ThilinaRajapakse** [comment]
Check how much GPU memory is in use and try lowering the `eval_batch_size` in case the evaluation is exceeding the available GPU memory. Rerunning cells inside Jupyter tends to cause memory issues, as it doesn't release memory properly.

Also, it might be better to get the labels from the entire df rather than just the train df on the off chance that the dev df contains labels that are not in train.
```
num_labels=len(df_bert['labels'].unique())
```

**Comment by sidSingla** [comment]
@ThilinaRajapakse 
Hi,
Just letting you know that I faced the same error 'RuntimeError: CUDA error: device-side assert triggered' today with the latest version. I tried a lot of things to overcome it, but in the end I tried a lower version of singletransformers(0.20.0) and it worked. I was earlier trying on the latest version(0.22.0). 
I was trying this example: https://towardsdatascience.com/simple-transformers-introducing-the-easiest-bert-roberta-xlnet-and-xlm-library-58bf8c59b2a3

Thanks a lot for the great wrapper :)

**Comment by ThilinaRajapakse** [comment]
Do you mind running with `use_cuda=False` and showing me the error?

**Comment by DmLitov4** [comment]
> @ThilinaRajapakse
> Hi,
> Just letting you know that I faced the same error 'RuntimeError: CUDA error: device-side assert triggered' today with the latest version(0.22.0). After trying a lot of things to overcome it, I tried a lower version of simpletransformers(0.20.0) in the end and it worked.
> I was trying this example: https://towardsdatascience.com/simple-transformers-introducing-the-easiest-bert-roberta-xlnet-and-xlm-library-58bf8c59b2a3
> 
> Thanks a lot for the great wrapper :)

I had the same issue. Switching to 0.20.0 version helped me, thanks!

**Comment by ThilinaRajapakse** [comment]
I can't replicate the issue guys. Is it only happening with the code from the Medium article?

**Comment by DmLitov4** [comment]
> I can't replicate the issue guys. Is it only happening with the code from the Medium article?

I've used my own dataset (multi-class classification). After downgrading on 0.20.0 (and then to 0.21.4) it works fine on GPU. Sometimes I have to restart the kernel but still it's OK.
But now I have another issue simialr to #267 . During training and validating the loss score is OK, but when I try to predict, for different texts I got almost same probabilities. I've balanced the dataset, tried different hyperparameters, etc. But nothing helps.

`myargs = {
    "use_early_stopping": True,
    "early_stopping_patience": 2,
    "early_stopping_delta": 0,
    "early_stopping_metric": "eval_loss",
    "early_stopping_metric_minimize": True,
    'fp16': False,
    'num_train_epochs': 3,
    "learning_rate": 0.0005,
    "max_seq_length": 128,
    "overwrite_output_dir": True,
}
model = ClassificationModel('xlnet', model_name='xlnet-base-cased', num_labels=4, use_cuda=True, cuda_device=0, args=myargs) `

**Comment by sidSingla** [comment]
@ThilinaRajapakse I followed the steps on medium article, though dataset was different. Will try to reproduce it again.

@DmLitov4 I was also running into same issues of same probabilities while predicting. I was iterating over the test dataset, and the probability weights and outputs for each of the data row was same. But then I passed the test set as whole to model.predict and then got different output values. Maybe you are also doing the same.
@ThilinaRajapakse can maybe comment on this.

Thanks.

**Comment by ThilinaRajapakse** [comment]
The issue with similar predictions was a bug in some of the older versions. I'm guessing the versions you guys downgraded to was afflicted with it as well.

**Comment by DmLitov4** [comment]
> The issue with similar predictions was a bug in some of the older versions. I'm guessing the versions you guys downgraded to was afflicted with it as well.

Yes, I've heard about this issue. But I've used 0.21.4 version (without that problem) and it seems like the problem was hiding behind hyperparameters of my model. I've changed it a lot, retrained my model and now it words fine with both XLNet and Bert.
Thank you so much your hard work, it helps us a lot.
One more question: is it possible to get f1-score for each class and not macro/micro average f1 during eval_model? Is there a code snippet for that? Thank you.
